geomstats/geometry/product_hpd_matrices_and_siegel_disks.py

```diff
@@ -0,0 +1,111 @@
+"""The ProductHPDMatricesAndSiegelDisks manifold.
+
+The ProductHPDMatricesAndSiegelDisks manifold is defined as a
+product manifold of the HPD manifold and (n-1) Siegel disks.
+The HPD Siegel disks product has a product metric.
+The product metric on the HPD Siegel disks product space is the usual
+HPD matrices affine-invariant metric (with power affine parameter equal 1)
+and Siegel metrics multiplied by constants.
+This product manifold can be used to represent Block-Toeplitz HPD matrices.
+
+
+Lead author: Yann Cabanes.
+
+References
+----------
+.. [Cabanes2022] Yann Cabanes. Multidimensional complex stationary
+    centered Gaussian autoregressive time series machine learning
+    in Poincaré and Siegel disks: application for audio and radar
+    clutter classification, PhD thesis, 2022
+.. [Cabanes2021] Yann Cabanes and Frank Nielsen.
+    New theoreticla tools in the Siegel space for vectorial
+    autoregressive data classification,
+    Geometric Science of Information, 2021.
+    https://franknielsen.github.io/IG/GSI2021-SiegelLogExpClassification.pdf
+.. [JV2016] B. Jeuris and R. Vandebril. The Kahler mean of Block-Toeplitz
+    matrices with Toeplitz structured blocks, 2016.
+    https://epubs.siam.org/doi/pdf/10.1137/15M102112X
+"""
+
+from geomstats.geometry.hpd_matrices import HPDAffineMetric, HPDMatrices
+from geomstats.geometry.product_manifold import ProductManifold
+from geomstats.geometry.product_riemannian_metric import ProductRiemannianMetric  # NOQA
+from geomstats.geometry.siegel import Siegel, SiegelMetric
+
+
+class ProductHPDMatricesAndSiegelDisks(ProductManifold):
+    """Class for the ProductHPDMatricesAndSiegelDisks manifold.
+
+    The HPD and Siegel product manifold is a direct product of the HPD manifold
+    and (n-1) Siegel disks. Each manifold of the product is a square matrix
+    manifold of the same dimension.
+
+    Parameters
+    ----------
+    n_manifolds : int
+        Number of manifolds of the product.
+    n : int
+        Size of the matrices.
+    """
+
+    def __init__(self, n_manifolds, n, **kwargs):
+        hpd_matrices = HPDMatrices(n=n)
+        hpd_matrices.metric = HPDAffineMetric(n=n, scale=n_manifolds**0.5)
+        siegel_disk = Siegel(n=n)
+        list_manifolds = [hpd_matrices] + (n_manifolds - 1) * [siegel_disk]
+        super(ProductHPDMatricesAndSiegelDisks, self).__init__(
+            factors=list_manifolds, **kwargs
+        )
+        self.shape = (n_manifolds, n, n)
```

Comment:
this is hacking `ProductManifold`.

We need to refactor `ProductManifold` to handle this situation "natively". Can you please open a PR only with the changes to product manifold, so we can focus on it? (I'll help you moving fast with it, and I also would like to bring @johnharveymath to the discussion, since he worked a lot on its refactoring.)

We can test the behavior you want by e.g. doing a product manifold with 2 `SPDMatrices` or any matrices space combination with same point shape.

Reply:
I am happy to get in on this. I can contribute on Friday. Just for clarification - the idea here is that the factors coming in are of shape (n, n) and we then want the product manifold to have a shape like (k, n, n)? I approve of this idea as it reflects best how we would mathematically think about the manifold.

Reply:
Exactly @johnharveymath. We can move this on in a different PR.

Do you have in mind the required changes to implement this? (I have some ideas in case.) I was thinking that maybe we should replace `default_point_type` by something more generic, like `point_ndim`, so it seems more naturally to make this change (this variable basically describes `len(shape)`. But probably this is not even required.

Let me know you're thoughts.

Reply:
Yes, this may be a good idea, but I don't think it's necessary for this particular issue. So far what I've done is add a `default_point_type` of `'other'` for `len(shape)>2`. The `ProductManifold` will try to choose the best shape for the user, which will be `(n_factors, factors.shape)` if they all have the same shape. That's working fine. Now I have to make sure nothing is broken in any other classes. by having this new type. Then I need to to just run through all the methods of `ProductManifold` to make them work with this shape, but I don't think that will be a problem at all.